### ISSUE TITLE
Bring back "failed build caching"

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nix#4127]
Removed in https://github.com/NixOS/nix/commit/8cffec84859cec8b610a2a22ab0c4d462a9351ff. This was raised a couple of times:

- https://twitter.com/acid2/status/1314143671911100421
- https://github.com/NixOS/nix/commit/8cffec84859cec8b610a2a22ab0c4d462a9351ff#commitcomment-29706519
- https://discourse.nixos.org/t/behavior-of-undeclared-broken-packages-is-confusing/2123/6?u=domenkozar
- https://discourse.nixos.org/t/why-do-nix-try-to-realise-a-derivation-that-already-failed/6026/4

In short, when you use a CI and you have builds failing, it's really wasteful to rebuild those on each commit.

**Comment by SuperSandro2000** [comment]
I would really appreciate that for nixpkgs-review to save build time.

**Comment by expipiplus1** [comment]
It's also amazingly useful when bisecting history!

**Comment by expipiplus1** [comment]
Had to waste CPU cycles bisecting something today, what's the word on this?

**Comment by domenkozar** [comment]
I'd suggest someone open a PR.

**Comment by kevincox** [comment]
It might make sense to be opt-in. So by default the behaviour doesn't change but you can pass `--use-fail-cache` or similar to opt-in. We can then argue about what the default should be.

**Comment by ocharles** [comment]
I would still like this

**Comment by theoparis** [comment]
I would love to see this implemented, so I don't have to rebuild my entire LLVM package when it fails 👍🏼

**Comment by fricklerhandwerk** [comment]
Triaged in Nix team meeting:

We're interested in such a feature, but the problem and solution space need to be explored more thoroughly. We welcome a design proposal.

<details><summary>Complete discussion</summary>

- @thufschmitt: what was the motivation for removing it?
    - @edolstra: it's error prone. if you had transient failures, those would also be cached
    - @thufschmitt: could it be smart enough to avoid caching known problems?
        - @edolstra: there already was some hackery to detect things like low disk space conditions
            - but it's a pretty ugly feature altogether
    - @thufschmitt: can we make it opt-in?
        - @edolstra: it's still code we'd have to maintain
        - @thufschmitt: it would make lots of sense for CIs
- @edolstra: Hydra doesn't need it, and it's questionable if other CIs really need it
    - it wouldn't even work for distributed builds as it's only implemented for the local store
        - there is no negative binary cache entry
        - @thufschmitt: but that's not a fundamental, conceptual limitation
- @thufschmitt: use cases mentioned in the discussion
    - CI
    - `nixpkgs-reviews`
    - bisecting history
- @edolstra: it could be handled by some pre-build hook
    - then it would be a policy for the user
    - @fricklerhandwerk: then we'd need a failed-build hook
- @fricklerhandwerk: this is a fundamental property of how Nix works, so we should take a closer look
    - @edolstra @thufschmitt: we probably won't implement it ourselves given limited time and other priorities, otherwise it would have happened already


</details>
